docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+    dist_ckpt_root.mkdir(exist_ok=True, parents=True)
+
+    torch.distributed.init_process_group()
+    world_size = torch.distributed.get_world_size()
+    rank = torch.distributed.get_rank()
+
+    # Local tensor to save
+    assert 128 % world_size == 0
+    num_elems_per_rank = 128 // world_size
+    local_ten = torch.arange(start=num_elems_per_rank * rank,
+                             end=num_elems_per_rank * (rank + 1))
+
+    # Native checkpoint save
+    state_dict = {
+        'weight': local_ten
+    }
+    torch.save(state_dict, native_ckpt_root / f'ckpt_{rank}.pt')
+
+    # Distributed checkpoint save
+    sharded_state_dict = {
+        'weight': dist_checkpointing.ShardedTensor.from_rank_offsets('weight', local_ten, (0, rank, world_size))
+    }
+    dist_checkpointing.save(sharded_state_dict, dist_ckpt_root)
+
+During load the distributed checkpoint can be easily read even if the job size changes (contrary to native checkpoints that require the same number of ranks).
+The main difference wrt. `torch.load` is that the user has to provide the definition of the sharded state dict that needs to be loaded.
+
+.. code-block:: python
+    from pathlib import Path
+
+    import torch
+
+    from megatron.core import dist_checkpointing
+
+    ckpt_root = Path('/tmp/checkpoints')
+    dist_ckpt_root = ckpt_root / 'dist_ckpt'
+
+    torch.distributed.init_process_group()
+    world_size = torch.distributed.get_world_size()
+    rank = torch.distributed.get_rank()
+    assert 128 % world_size == 0
+    num_elems_per_rank = 128 // world_size
+
+    # Local tensor to load
+    local_ten = torch.empty(num_elems_per_rank)
+    sharded_state_dict = {
+        'weight': dist_checkpointing.ShardedTensor.from_rank_offsets('weight', local_ten, (0, rank, world_size))
+    }
+    loaded_state_dict = dist_checkpointing.load(sharded_state_dict, dist_ckpt_root)
+    expected_local_ten = torch.arange(start=num_elems_per_rank * rank, end=num_elems_per_rank * (rank + 1))
+    assert torch.all(loaded_state_dict['weight'] == expected_local_ten)
+
+    # With torch.save and torch.load, we would have to load all files that contain
+    # parts of the desired tensor in new configuration and concatenate appropriate fragments.
+    # For some distributed checkpoint backends this is actually what happens underneath.
+
+
+Supported entities
+==================
+The distributed checkpointing library supports saving and loading of different objects in different configurations.
+
+A sharded state dict is a (possibly nested) Python dictionary or list with the following elements:
+
+1. ShardedBase
+    a. ShardedTensor
+    #. ShardedObject
+    #. ShardedTensorFactory
+#. LocalNonpersitentObject
+#. Arbitrary object
+
+
+ShardedBase
+-----------
+Base class for expressing any kind of sharding.
+Each sharded entity must be uniquely identified by its `key`, carry some `data` to be saved or loaded and define `replica_id` which helps identify data redundancy.
```

Comment:
Each sharded entity must be uniquely identified by its `key`, carry some `data` to be saved or loaded, and define `replica_id` which helps identify data redundancy.